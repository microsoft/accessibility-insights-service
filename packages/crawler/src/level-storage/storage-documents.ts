// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { SummaryScanResults } from 'accessibility-insights-report';

export type ItemType = 'failedScanResult' | 'passedScanResult' | 'browserError' | 'runError' | 'scanMetadata';

export interface DataBaseKey {
    type: ItemType;
    key: string;
}

export interface PageError {
    url: string;
    error: string;
}

export interface ScanResults {
    summaryScanResults: SummaryScanResults;
    errors: PageError[];
    scanMetadata: ScanMetadata;
}

export interface ScanMetadata {
    baseUrl: string;
    basePageTitle?: string;
    userAgent?: string;
}
