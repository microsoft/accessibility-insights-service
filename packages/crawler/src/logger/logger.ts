// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import { injectable } from 'inversify';
import { ApifyConsoleLoggerClient } from './apify-console-logger-client';

export enum LogLevel {
    Info,
    Warn,
    Verbose,
    Error,
}

export interface LoggerProperties {
    [name: string]: string;
}

export interface LoggerClient {
    log(message: string, logLevel: LogLevel, properties?: LoggerProperties): void;
    setCommonProperties(properties: LoggerProperties): void;
}

@injectable()
export class Logger {
    constructor(private readonly loggerClients: LoggerClient[] = [new ApifyConsoleLoggerClient()]) {}

    public setCommonProperties(properties: LoggerProperties): void {
        this.invokeLoggerClients((client) => client.setCommonProperties(properties));
    }

    public log(message: string, logLevel: LogLevel, properties?: LoggerProperties): void {
        this.invokeLoggerClients((client) => client.log(message, logLevel, properties));
    }

    public logInfo(message: string, properties?: LoggerProperties): void {
        this.log(message, LogLevel.Info, properties);
    }

    public logVerbose(message: string, properties?: LoggerProperties): void {
        this.log(message, LogLevel.Verbose, properties);
    }

    public logWarn(message: string, properties?: LoggerProperties): void {
        this.log(message, LogLevel.Warn, properties);
    }

    public logError(message: string, properties?: LoggerProperties): void {
        this.log(message, LogLevel.Error, properties);
    }

    private invokeLoggerClients(action: (loggerClient: LoggerClient) => void): void {
        this.loggerClients.map(action);
    }
}
