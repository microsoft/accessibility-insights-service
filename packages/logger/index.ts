export { Logger, LogLevel } from './logger';
export { BaseEntryPoint } from './base-entry-point';
export { loggerTypes } from './logger-types';
export { registerLoggerToContainer } from './register-logger-to-container';
export { BaseTelemetryProperties } from './base-telemetry-properties';
