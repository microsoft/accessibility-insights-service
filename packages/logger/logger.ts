import { inject, injectable } from 'inversify';

import * as appInsights from 'applicationinsights';
import { loggerTypes } from './logger-types';

@injectable()
export class Logger {
    private initialized: boolean = false;

    constructor(
        @inject(loggerTypes.AppInsights) private readonly appInsightsObject: typeof appInsights,
        @inject(loggerTypes.Process) private readonly currentProcess: typeof process,
    ) {}

    public setup(baseProperties: { [key: string]: string }): void {
        if (this.initialized === true) {
            return;
        }

        this.appInsightsObject
            .setup()
            .setAutoCollectConsole(false)
            .setAutoCollectExceptions(true)
            .setAutoCollectRequests(false);

        // this should be set after calling setup
        this.appInsightsObject.defaultClient.commonProperties = {
            batchPoolId: this.currentProcess.env.AZ_BATCH_POOL_ID,
            batchJobId: this.currentProcess.env.AZ_BATCH_JOB_ID,
            batchTaskId: this.currentProcess.env.AZ_BATCH_TASK_ID,
            batchNodeId: this.currentProcess.env.AZ_BATCH_NODE_ID,
            ...baseProperties,
        };

        this.appInsightsObject.start();
        this.initialized = true;
    }

    public trackMetric(name: string, value: number = 1): void {
        this.ensureInitialized();

        this.appInsightsObject.defaultClient.trackMetric({ name: name, value: value });
    }

    public trackEvent(name: string, properties?: { [name: string]: string }): void {
        this.ensureInitialized();

        this.appInsightsObject.defaultClient.trackEvent({ name: name, properties: properties });
    }

    public trackTrace(message: string, severity: appInsights.Contracts.SeverityLevel, properties?: { [name: string]: string }): void {
        this.ensureInitialized();

        this.appInsightsObject.defaultClient.trackTrace({ message: message, severity: severity, properties: properties });
    }

    public trackInfoTrace(message: string, properties?: { [name: string]: string }): void {
        this.trackTrace(message, appInsights.Contracts.SeverityLevel.Information, properties);
    }

    public flush(): void {
        this.appInsightsObject.defaultClient.flush();
    }

    private ensureInitialized(): void {
        if (this.initialized === true) {
            return;
        }

        throw new Error('Telemetry client not setup');
    }
}
