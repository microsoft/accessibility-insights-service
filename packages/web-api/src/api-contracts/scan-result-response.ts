// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
export declare type LinkType = 'self';
export declare type ReportFormat = 'sarif';
export declare type ScanState = 'unknown' | 'pass' | 'fail';
export declare type RunState = 'not found' | 'accepted' | 'queued' | 'running' | 'completed' | 'failed';

export interface ScanResultErrorResponse {
    scanId: string;
    error: string;
}

export type ScanResultResponse =
    | {
          scanId: string;
          url: string;
          scanResult?: ScanResult;
          reports?: ScanReport[];
          run: ScanRun;
      }
    | ScanResultErrorResponse;

export interface ScanResult {
    state: ScanState;
    issueCount?: number;
}

export interface ScanReport {
    reportId: string;
    format: ReportFormat;
    links?: Links;
}

export interface ScanRun {
    state: RunState;
    timestamp?: string;
    error?: string;
}

export interface Links {
    rel: LinkType;
    href: string;
}
