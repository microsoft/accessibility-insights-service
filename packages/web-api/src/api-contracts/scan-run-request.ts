// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
export interface ScanRunRequest {
    url: string;

    /**
     * Priority values can range from -1000 to 1000, with -1000 being the lowest priority and 1000 being the highest priority.
     * The default value is 0.
     */
    priority?: number;
}
