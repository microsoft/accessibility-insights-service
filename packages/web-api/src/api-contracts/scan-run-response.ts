// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { WebApiError } from 'service-library';

export interface ScanRunResponse {
    scanId?: string;
    url: string;
    error?: WebApiError;
}
