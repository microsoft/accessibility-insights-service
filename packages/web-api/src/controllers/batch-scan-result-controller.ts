// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { GuidGenerator, ServiceConfiguration } from 'common';
import { inject, injectable } from 'inversify';
import { isEmpty } from 'lodash';
import { Logger } from 'logger';
import { OnDemandPageScanRunResultProvider } from 'service-library';
import { ScanResultResponse } from './../api-contracts/scan-result-response';

import { ScanBatchRequest } from './../api-contracts/scan-batch-request';
import { BaseScanResultController } from './base-scan-result-controller';

@injectable()
export class BatchScanResultController extends BaseScanResultController {
    public readonly apiVersion = '1.0';
    public readonly apiName = 'wep-api-get-scans';

    public constructor(
        @inject(OnDemandPageScanRunResultProvider) protected readonly onDemandPageScanRunResultProvider: OnDemandPageScanRunResultProvider,
        @inject(GuidGenerator) protected readonly guidGenerator: GuidGenerator,
        @inject(ServiceConfiguration) protected readonly serviceConfig: ServiceConfiguration,
        @inject(Logger) protected readonly logger: Logger,
    ) {
        super();
    }

    public async handleRequest(): Promise<void> {
        const payload = this.tryGetPayload<ScanBatchRequest[]>();
        if (isEmpty(payload)) {
            this.context.res = {
                status: 422, // Unprocessable Entity,
            };

            return;
        }

        const scanIds = payload.map(request => request.scanId);
        const responseBody: ScanResultResponse[] = [];
        const scanIdsToQuery: string[] = [];

        for (const scanId of scanIds) {
            if (!this.isScanIdValid(scanId)) {
                responseBody.push(this.getInvalidRequestResponse(scanId));
                continue;
            }
            const isRequestMadeTooSoon = await this.isRequestMadeTooSoon(scanId);
            if (isRequestMadeTooSoon === true) {
                responseBody.push(this.getTooSoonRequestResponse(scanId));
            } else {
                scanIdsToQuery.push(scanId);
            }
        }

        const scanResultItemMap = await this.getScanResultMapKeyByScanId(scanIdsToQuery);

        scanIdsToQuery.forEach(scanId => {
            if (isEmpty(scanResultItemMap[scanId])) {
                responseBody.push(this.get404Response(scanId));
            } else {
                responseBody.push(this.getResponseFromDbDocument(scanResultItemMap[scanId]));
            }
        });

        this.context.res = {
            status: 200,
            body: responseBody,
        };

        this.logger.logInfo('batch scan result fetched');
    }

    // tslint:disable-next-line: no-empty
    protected handleInvalidRequest(): void {}
}
