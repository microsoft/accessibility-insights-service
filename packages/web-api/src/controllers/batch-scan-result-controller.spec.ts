// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import 'reflect-metadata';

import { Context } from '@azure/functions';
import { GuidGenerator, RestApiConfig, ServiceConfiguration } from 'common';
import { Logger } from 'logger';
import { OnDemandPageScanRunResultProvider } from 'service-library';
import { ItemType, OnDemandPageScanResult } from 'storage-documents';
import { IMock, It, Mock, Times } from 'typemoq';

import { ScanBatchRequest } from './../api-contracts/scan-batch-request';
import { ScanResultResponse } from './../api-contracts/scan-result-response';
import { BatchScanResultController } from './batch-scan-result-controller';

describe(BatchScanResultController, () => {
    let batchScanResultController: BatchScanResultController;
    let context: Context;
    let onDemandPageScanRunResultProviderMock: IMock<OnDemandPageScanRunResultProvider>;
    let serviceConfigurationMock: IMock<ServiceConfiguration>;
    let loggerMock: IMock<Logger>;
    let guidGeneratorMock: IMock<GuidGenerator>;
    const validScanId = 'valid-scan-id';
    const notFoundScanId = 'not-found-scan-id';
    const invalidScanId = 'invalid-scan-id';
    const requestedTooSoonScanId = 'requested-too-soon-scan-id';
    const batchRequestBody: ScanBatchRequest[] = [
        { scanId: validScanId },
        { scanId: notFoundScanId },
        { scanId: invalidScanId },
        { scanId: requestedTooSoonScanId },
    ];
    const scanFetchedResponse: OnDemandPageScanResult = {
        id: validScanId,
        partitionKey: 'pk',
        url: 'url',
        run: {
            state: 'running',
        },
        priority: 1,
        itemType: ItemType.onDemandPageScanRunResult,
        reports: [
            {
                reportId: 'report-id',
                format: 'sarif',
                href: 'href',
            },
        ],
    };

    beforeEach(() => {
        context = <Context>(<unknown>{
            req: {
                method: 'POST',
                headers: {},
                rawBody: `[]`,
                query: {},
            },
        });
        context.req.query['api-version'] = '1.0';
        context.req.headers['content-type'] = 'application/json';
        onDemandPageScanRunResultProviderMock = Mock.ofType<OnDemandPageScanRunResultProvider>();
        onDemandPageScanRunResultProviderMock
            // tslint:disable-next-line: no-unsafe-any
            .setup(async o => o.readScanRuns(It.isAny()))
            .returns(async () => Promise.resolve([scanFetchedResponse]));

        guidGeneratorMock = Mock.ofType(GuidGenerator);
        guidGeneratorMock
            .setup(gm => gm.isValidV6Guid(It.isAnyString()))
            .returns(id => {
                return id !== invalidScanId;
            });

        serviceConfigurationMock = Mock.ofType<ServiceConfiguration>();
        serviceConfigurationMock
            .setup(async s => s.getConfigValue('restApiConfig'))
            .returns(async () => {
                // tslint:disable-next-line: no-object-literal-type-assertion
                return {
                    maxScanRequestBatchCount: 2,
                    minimumWaitTimeforScanResultQueryInSeconds: 300,
                } as RestApiConfig;
            });

        loggerMock = Mock.ofType<Logger>();
    });

    function createScanResultController(contextReq: Context): BatchScanResultController {
        const controller = new BatchScanResultController(
            onDemandPageScanRunResultProviderMock.object,
            guidGeneratorMock.object,
            serviceConfigurationMock.object,
            loggerMock.object,
        );
        controller.context = contextReq;

        return controller;
    }

    function setupGetGuidTimestamp(scanId: string, time: Date): void {
        guidGeneratorMock
            .setup(gm => gm.getGuidTimestamp(scanId))
            .returns(() => time)
            .verifiable(Times.once());
    }

    describe('handleRequest', () => {
        it('should return 422 if request body is empty array', async () => {
            context.req.rawBody = undefined;
            batchScanResultController = createScanResultController(context);

            await batchScanResultController.handleRequest();

            expect(context.res.status).toEqual(422);
        });

        it('should return different response for different kind of scanIds', async () => {
            context.req.rawBody = JSON.stringify(batchRequestBody);
            batchScanResultController = createScanResultController(context);
            const requestTooSoonTimeStamp = new Date();
            requestTooSoonTimeStamp.setFullYear(requestTooSoonTimeStamp.getFullYear() + 1);
            const validTimeStamp = new Date(0);

            setupGetGuidTimestamp(validScanId, validTimeStamp);
            setupGetGuidTimestamp(requestedTooSoonScanId, requestTooSoonTimeStamp);
            setupGetGuidTimestamp(notFoundScanId, validTimeStamp);

            await batchScanResultController.handleRequest();

            guidGeneratorMock.verifyAll();
            expect(context.res.status).toEqual(200);
            expect(context.res.body).toMatchSnapshot();
        });
    });
});
