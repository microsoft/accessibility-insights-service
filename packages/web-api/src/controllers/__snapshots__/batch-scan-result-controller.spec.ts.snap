// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BatchScanResultController handleRequest should return different response for different kind of scanIds 1`] = `
Array [
  Object {
    "error": Object {
      "code": "InvalidResourceId",
      "codeId": 4002,
      "message": "The resource ID is not valid.",
    },
    "scanId": "invalid-scan-id",
  },
  Object {
    "run": Object {
      "state": "pending",
    },
    "scanId": "requested-too-soon-scan-id",
    "url": undefined,
  },
  Object {
    "run": Object {
      "state": "running",
    },
    "scanId": "valid-scan-id",
    "url": "url",
  },
  Object {
    "error": Object {
      "code": "ResourceNotFound",
      "codeId": 4001,
      "message": "The specified resource does not exist.",
    },
    "scanId": "not-found-scan-id",
  },
]
`;
